File: tools/serve.py
#!/usr/bin/env python3
"""Serve the app on localhost so it can be used without internet hosting.

Opening index.html by double-clicking does not work: the app is built from ES
modules, which a browser refuses to load over file://, and it fetches the blank
PDF templates, which file:// also blocks. Both need a real HTTP origin, so this
serves one.

Bound to 127.0.0.1 on purpose. The default for http.server is every interface,
which would put a form holding someone's SSN on the local network.
"""

import functools
import http.server
import sys
import threading
import webbrowser
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
FIRST_PORT = 8000
TRIES = 20


class Server(http.server.ThreadingHTTPServer):
    # ThreadingHTTPServer, not a bare TCPServer, for two reasons. It sets
    # allow_reuse_address, without which stopping and restarting inside the
    # TIME_WAIT window is refused and the retry loop below silently moves the
    # app to a different port -- so the URL the user was told changes. And it
    # serves in parallel, which a browser opening six connections for the
    # modules and a 1.3 MB template needs.
    daemon_threads = True


class Handler(http.server.SimpleHTTPRequestHandler):
    def end_headers(self):
        # Without this a stale copy survives a `git pull`, and the user is
        # debugging a version they no longer have.
        self.send_header("Cache-Control", "no-store")
        super().end_headers()

    def log_message(self, fmt, *args):
        # 404s matter (a missing template is worth seeing); 200s are noise.
        if not str(args[1] if len(args) > 1 else "").startswith("2"):
            super().log_message(fmt, *args)


def main():
    handler = functools.partial(Handler, directory=str(ROOT))

    for port in range(FIRST_PORT, FIRST_PORT + TRIES):
        try:
            with Server(("127.0.0.1", port), handler) as httpd:
                url = f"http://localhost:{port}/"
                print()
                print("  The assistant is running at:")
                print(f"      {url}")
                print()
                print("  Leave this window open while you use it.")
                print("  Press Control-C here when you are finished.")
                print()
                threading.Timer(0.5, webbrowser.open, args=[url]).start()
                try:
                    httpd.serve_forever()
                except KeyboardInterrupt:
                    print("\n  Stopped. You can close this window.")
                return 0
        except OSError:
            continue  # port taken, try the next one

    print(f"Could not find a free port between {FIRST_PORT} and "
          f"{FIRST_PORT + TRIES - 1}.", file=sys.stderr)
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: tools/serve.mjs
// Serve the app on localhost, using only Node's standard library.
//
// This is the fallback for machines with Node but no Python. It mirrors
// tools/serve.py exactly: same ports, same 127.0.0.1-only binding, same
// browser launch. See that file for why file:// is not an option.

import { createServer } from 'node:http';
import { spawn } from 'node:child_process';
import { createReadStream } from 'node:fs';
import { stat } from 'node:fs/promises';
import { extname, join, resolve, sep } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = resolve(fileURLToPath(new URL('..', import.meta.url)));
const FIRST_PORT = 8000;
const TRIES = 20;

// A wrong type here is not cosmetic: a browser refuses a module served as
// anything but JavaScript, and the whole app is modules.
const TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.pdf': 'application/pdf',
  '.mp3': 'audio/mpeg',
  '.wav': 'audio/wav',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.ico': 'image/vnd.microsoft.icon',
  '.woff2': 'font/woff2'
};

function openBrowser(url) {
  const cmd = process.platform === 'darwin' ? 'open'
    : process.platform === 'win32' ? 'cmd'
      : 'xdg-open';
  const args = process.platform === 'win32' ? ['/c', 'start', '', url] : [url];
  try {
    spawn(cmd, args, { stdio: 'ignore', detached: true }).unref();
  } catch {
    /* the URL is printed either way */
  }
}

const server = createServer(async (req, res) => {
  const send = (code, body) => {
    res.writeHead(code, { 'Content-Type': 'text/plain; charset=utf-8' });
    res.end(body);
  };

  let pathname;
  try {
    ({ pathname } = new URL(req.url, 'http://localhost'));
  } catch {
    return send(400, 'Bad request');
  }

  let file = resolve(join(ROOT, decodeURIComponent(pathname)));

  // Keep the server inside the project directory. Without this, a crafted
  // path walks up into the user's home folder.
  if (file !== ROOT && !file.startsWith(ROOT + sep)) return send(403, 'Forbidden');

  try {
    if ((await stat(file)).isDirectory()) file = join(file, 'index.html');
  } catch {
    return send(404, 'Not found');
  }

  try {
    const info = await stat(file);
    res.writeHead(200, {
      'Content-Type': TYPES[extname(file).toLowerCase()] ?? 'application/octet-stream',
      'Content-Length': info.size,
      'Cache-Control': 'no-store'
    });
    createReadStream(file).pipe(res);
  } catch {
    send(404, 'Not found');
  }
});

let port = FIRST_PORT;

server.on('error', err => {
  if (err.code === 'EADDRINUSE' && port < FIRST_PORT + TRIES - 1) {
    server.listen(++port, '127.0.0.1');   // port taken, try the next one
    return;
  }
  console.error(`Could not start the server: ${err.message}`);
  process.exit(1);
});

server.on('listening', () => {
  const url = `http://localhost:${port}/`;
  console.log();
  console.log('  The assistant is running at:');
  console.log(`      ${url}`);
  console.log();
  console.log('  Leave this window open while you use it.');
  console.log('  Press Control-C here when you are finished.');
  console.log();
  setTimeout(() => openBrowser(url), 500);
});

process.on('SIGINT', () => {
  console.log('\n  Stopped. You can close this window.');
  process.exit(0);
});

server.listen(port, '127.0.0.1');
